### PR TITLE
Simplify FileCacheStorage module

### PR DESCRIPTION
**Pull request by CyrilleContamine**
In ICacheStorage I only kept : Add (previously named Set), Remove (previously Delete), Get 

I also changed some public methods to make them private : Has(), MatchesVersion(), DeleteAll() and removed SaveCacheStorageFile()

I also renamed TryLoadCacheEntries to DeserializeCacheEntries to make it more clear that we are deserializing stuff

I also changed the variable Version, CacheEntries to make them private and used [SerializeField] to make it work with JsonUtility (I've checked on Display3D and it actually works, but I needed to add [FormerlySerializedAs("CacheEntries")] for the deserialization to work of course). Note : I could have changed to use Json.Net as well

------

NOTE 31/01/2022 : After many time spent on modifying part of the code and removing stuff (no more versioning for both file in cache and for the cache file itself), I think having the FileCacheEntry does not bring value anymore, so I removed it and everything related. When we need to fetch a data from the cache I simplify check on the FS to get value from it, no more intermediary to tell us where to look at at since we know we should be looking at a specific file located at _cachePath/id.

### CODE REVIEW

File: 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs [hunk 1]
```diff
@@ -128,8 +132,7 @@ private List<FileCacheEntry> GetExpiredEntries()
 
 	private void UpdateUseDate( string id )
 	{
-		if( _cacheEntries.ContainsKey( id ) )
```
Comment (Jeantkt) [inline 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs:131]: why removing this condition?

Comment (Jeantkt) [reply 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs:131]: ah ok when using this function, we know that the id exists

Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs:131]: Because the method says it will update the use date and its private so we should make sure the entry exist before calling this method (it's actually done). I already removed the check from the Has method last time but got no comment on it so I did the same here.

Comment (Jeantkt) [reply 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs:131]: ok for me, I wanted to close this thread but I don't know how to do it

Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs:131]: Thx , I'll close it :)

File: 3D/senior-csharp-test/Assets/Cache/Test/FileCacheStorageTest.cs [hunk 2]
```diff
@@ -1,7 +1,6 @@
 using NUnit.Framework;
 using Zenject;
 
-
 [ TestFixture ]
 public class FileCacheStorageTest : ZenjectUnitTestFixture
```
Comment (Jeantkt) [inline 3D/senior-csharp-test/Assets/Cache/Test/FileCacheStorageTest.cs]: if we add a Sample test, we could add an empty Setup as well? or do you think it's not necessary?

Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/Test/FileCacheStorageTest.cs]: I think we discussed about not helping the user on the testing thing (that's why I remove also the hint), he should know already OR check by himself.

However I wonder if we need to give him zenject since we need to call the FileCacheStorage constructor ourselves (can't be injected like we do for other module)

Comment (Jeantkt) [reply 3D/senior-csharp-test/Assets/Cache/Test/FileCacheStorageTest.cs]: I guess we should not include it yes, it may be confusing to leave it

Comment (Jeantkt) [reply 3D/senior-csharp-test/Assets/Cache/Test/FileCacheStorageTest.cs]: but since we added a SampleTest, I was wondering what is the extend of having samples in the class

File: 3D/senior-csharp-test/Assets/Cache/ICacheStorage.cs [hunk 3]
```diff
@@ -1,9 +1,6 @@
 ﻿public interface ICacheStorage
 {
-	void Delete( string id );
-	void Set( byte[] data, string id, string version = null );
+	void Add( string id, byte[] data, string version );
+	void Remove( string id );
 	byte[] Get( string id );
-	bool Has( string id );
```
Comment (elicazaharieva) [inline 3D/senior-csharp-test/Assets/Cache/ICacheStorage.cs:6]: why did you make these private?

File: 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs [hunk 4]
```diff
@@ -52,16 +56,17 @@ private void InitCache()
...
 		}
 
-		if( TryLoadCacheEntries() )
+		if( DeserializeCacheEntries() )
 		{
-			CheckCacheVersion();
-			ClearExpiredEntries();
+			if( !IsCurrentVersion() )
+				DeleteAll();
+			ClearExpiredCacheEntries();
```
Comment (elicazaharieva) [inline 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs]: shouldn't this be in a `else`? 🤔

File: 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs [hunk 5]
```diff
@@ -1,22 +1,26 @@
...
 	private const string DATE_FORMAT = "yyMMdd";
 	private const float EXPIRING_TIME_DAYS = 60f;
+
 	private readonly string _cachePath;
-	private readonly Dictionary<string, FileCacheEntry> _cacheEntries = new Dictionary<string, FileCacheEntry>();
+	private readonly Dictionary<string, FileCacheEntry> _cacheEntryById = new Dictionary<string, FileCacheEntry>();
 	private readonly string _fileName;
 	private readonly IFileSystem _fileSystem;
 	private readonly ISerializer _serializer;
-	public List<FileCacheEntry> CacheEntries = new List<FileCacheEntry>();
```
Comment (elicazaharieva) [inline 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs:18]: you renamed this, so I'm not sure the cache deserializarion would work... or the json element would have a weird name... In any case we can't to the same rename in Display, it will break the existing cache 

maybe we should add a json property above. Or (I would find that even prettier), not have this list at all, but deserialize into a list directly (instead of `this`) and pass it to the method that fills in the dictionary

File: 3D/senior-csharp-test/Assets/Cache/ICacheStorage.cs (hunk 3 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/ICacheStorage.cs:6]: To reduce the number of things to be tested (I thought we wanted to reduce the tests so that the candidate does not take too much time on this).

But I think i should in fact remove everything related to the version (removed from CacheEntry constructor, from Add() parameters and remove MatchesVersion )

File: 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs (hunk 5 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs:18]: I've tried to explain in the MR description at the top what I've done here, I went the quick way for this MR but I can change and use Json.Net yes and use Property if you want (don't have that in JsonUtility). I I can try also what you propose (use JsonUtility.FromJson()), What do you prefer ?

File: 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs (hunk 4 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs]: Indeed (but that's like that on Display)

Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs]: done

File: 3D/senior-csharp-test/Assets/Cache/Test/FileCacheStorageTest.cs (hunk 2 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/Test/FileCacheStorageTest.cs]: Zenject + hints removed, Moq kept. Closing here

File: 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs (hunk 5 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs:18]: Finally I used Json.net (better to be able to deserialize to a list of interfaces so that the candidate can test) and I'm deserializing in a new list directly that I pass to the method that fills the dictionary from it

File: 3D/senior-csharp-test/Assets/Cache/ICacheStorage.cs (hunk 3 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/ICacheStorage.cs:6]: done removed everything related to version, for the cache entry and for the cache file as well

File: 3D/senior-csharp-test/Assets/Cache/CacheEntry.cs [hunk 6]
```diff
@@ -2,29 +2,24 @@
 using System.Collections.Generic;
 
 [ Serializable ]
-public class CacheEntry<T>
+public class CacheEntry<T> : ICacheEntry<T>
```
Comment (elicazaharieva) [inline 3D/senior-csharp-test/Assets/Cache/CacheEntry.cs]: I wonder why we need this intermediate class.. we could just declare `FileCacheEntry : ICacheEntry<string>  `

Am I missing something? Something related to the tests or serialization?

File: 3D/senior-csharp-test/README.md [hunk 7]
```diff
@@ -6,14 +6,14 @@ Unity3D 2018.3.12 ( https://download.unity3d.com/download_unity/8afd630d1f5b/Win
...
+Our tests are implemented using Moq and NUnit, you have all the necessary includes and assembly definitions provided. 
 
 ## Content
 
-We have provided a small module that implements saving files to a cache, and retrieving them. 
-The main class is FileCacheStorage, and this is the one we want to test. 
-We have provided the FileCacheStorageTest file to get you started, so all you need to do is write your test functions inside. 
-We aim at unit testing rather than integration or end-to-end tests.
-The naming convention we usually follow for unit tests is `MyFunction_DoThis_When_SomethingHappens`, allowing to clarify the intent of the test.
+We have provided a small module that implements saving files to a cache, and retrieving them. \
```
Comment (elicazaharieva) [inline 3D/senior-csharp-test/README.md]: hm, no saving any more. Maybe it's the right place to mention that the class is simplified and the cache save is not implemented?

File: 3D/senior-csharp-test/README.md [hunk 8]
```diff
@@ -6,14 +6,14 @@ Unity3D 2018.3.12 ( https://download.unity3d.com/download_unity/8afd630d1f5b/Win
...
+The main class is FileCacheStorage, and this is the one we want to test. \
+We have provided the FileCacheStorageTest file to get you started, so all you need to do is write your test functions inside. \
+We aim at unit testing rather than integration or end-to-end tests. \
+The naming convention we usually follow for unit tests is `MyFunction_DoThis_When_SomethingHappens`, allowing to clarify the intent of the test. \
```
Comment (elicazaharieva) [inline 3D/senior-csharp-test/README.md]:
```suggestion
The naming convention we usually follow for unit tests is `MyFunction_DoesThis_When_SomethingHappens`, allowing to clarify the intent of the test. \
```

File: 3D/senior-csharp-test/Assets/Cache/CacheEntry.cs (hunk 6 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/CacheEntry.cs]: Yes no need for it, removed

File: 3D/senior-csharp-test/README.md (hunk 7 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/README.md]: Indeed, I'll try to explain that it's the cache file save that is not implemented (because Add() still save cache files on the FS)

File: 3D/senior-csharp-test/README.md (hunk 8 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/README.md]: done

File: 3D/senior-csharp-test/README.md (hunk 7 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/README.md]: done

Comment (elicazaharieva) [reply 3D/senior-csharp-test/README.md]: true

Comment (CyrilleContamine) [reply 3D/senior-csharp-test/README.md]: I just wrote that the cache save is not implemented because finally I think it's still understandable

File: 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs [hunk 9]
```diff
@@ -52,218 +28,121 @@ private void InitCache()
...
+			{
+				Log.LogError( "Couldn't access file last access time at path = " + filePath + ".\n" + e );
+				continue;
+			}
+
+			TimeSpan span = dateNow - lastAccessDate;
 
 			if( span.TotalDays > EXPIRING_TIME_DAYS )
-				expiredEntries.Add( ce.Value );
+				expiredFilePaths.Add( filePath );
```
Comment (elicazaharieva) [inline 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs]: we could delete the file directly instead of filling in an intermediate list. Just a detail I guess

File: 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs [hunk 10]
```diff
@@ -52,218 +28,121 @@ private void InitCache()
...
-		else
-			DeleteAll();
+		ClearExpiredCacheFiles();
 	}
 
-	private bool TryLoadCacheEntries()
+	private void ClearExpiredCacheFiles()
 	{
-		string filePath = _cachePath + "/" + _fileName;
+		List<string> expiredfilePath = GetExpiredFilePaths();
```
Comment (elicazaharieva) [inline 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs]: #niptuck if you keep this method
```suggestion
		List<string> expiredfilePaths = GetExpiredFilePaths();
```

File: 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs (hunk 9 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs]: Indeed, I removed GetExpiredFilePaths and put his code directly inside ClearExpiredCacheFiles

File: 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs (hunk 10 above)
Comment (CyrilleContamine) [reply 3D/senior-csharp-test/Assets/Cache/FileCacheStorage.cs]: didn't keep it :)
